In [55]:
# Libraries
import numpy as np
import pandas as pd
import re

In [56]:
import sys
sys.path.append("..")
from src.cleaning import *

# Primera visualización rápida

Cargamos datos de formato CSV

In [57]:
attacks = pd.read_csv("../data/attacks.csv", encoding = "ISO-8859-1")

Visualizo las primeras y últimas filas para echar una ojeada a los datos.

In [58]:
attacks.head(2)

Case Number         Date    Year        Type Country        Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating     USA  California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked     USA     Georgia   

                         Location  Activity             Name Sex   ...  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely     F  ...   

      Species           Investigator or Source                     pdf  \
0  White shark                R. Collier, GSAF    2018.06.25-Wolfe.pdf   
1          NaN  K.McMurray, TrackingSharks.com  2018.06.18-McNeely.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  

[2 rows x 24 columns]

In [59]:
attacks.tail(2)

Case Number Date  Year Type Country Area Location Activity Name Sex   \
25721         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25722          xx  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

       ... Species  Investigator or Source  pdf href formula href  \
25721  ...      NaN                    NaN  NaN          NaN  NaN   
25722  ...      NaN                    NaN  NaN          NaN  NaN   

      Case Number.1 Case Number.2 original order Unnamed: 22 Unnamed: 23  
25721           NaN           NaN            NaN         NaN         NaN  
25722           NaN           NaN            NaN         NaN         NaN  

[2 rows x 24 columns]

Visualizo también una muestra aleatoria.

In [60]:
attacks.sample(2)

Case Number Date  Year Type Country Area Location Activity Name Sex   \
8629            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
14056         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

       ... Species  Investigator or Source  pdf href formula href  \
8629   ...      NaN                    NaN  NaN          NaN  NaN   
14056  ...      NaN                    NaN  NaN          NaN  NaN   

      Case Number.1 Case Number.2 original order Unnamed: 22 Unnamed: 23  
8629            NaN           NaN            NaN         NaN         NaN  
14056           NaN           NaN            NaN         NaN         NaN  

[2 rows x 24 columns]

# Mis HIPOTESIS.

1_ USA es el pais donde más accidentes de tiburón se producen.

2_ La mayor parte de los accidentes son no provocados.

3_ Fishing y Surfing; mayores causas de accidentes.

4_ Mueren más hombres que mujeres.

# Análisis exploratorio del DataFrame

Evaluar la forma, "info" y "Nan" del Data Frame. Para este caso no merece la pena hacer un "describe". 

In [61]:
attacks.shape

(25723, 24)

In [62]:
attacks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

In [63]:
attacks.isnull().sum()

Case Number               17021
Date                      19421
Year                      19423
Type                      19425
Country                   19471
Area                      19876
Location                  19961
Activity                  19965
Name                      19631
Sex                       19986
Age                       22252
Injury                    19449
Fatal (Y/N)               19960
Time                      22775
Species                   22259
Investigator or Source    19438
pdf                       19421
href formula              19422
href                      19421
Case Number.1             19421
Case Number.2             19421
original order            19414
Unnamed: 22               25722
Unnamed: 23               25721
dtype: int64

Hago comprbación de DATOS NUMËRICOS y CATEGÓRICOS

In [64]:
# DATOS NUMÉRICOS
numericas = attacks._get_numeric_data()
numericas.head(3)

Year  original order
0  2018.0          6303.0
1  2018.0          6302.0
2  2018.0          6301.0

In [65]:
# DATOS CATEGÓRICOS
attacks.select_dtypes(include="object").head(2)

Case Number         Date        Type Country        Area  \
0  2018.06.25  25-Jun-2018     Boating     USA  California   
1  2018.06.18  18-Jun-2018  Unprovoked     USA     Georgia   

                         Location  Activity             Name Sex  Age  ...  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe    F  57  ...   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely     F  11  ...   

            Time     Species           Investigator or Source  \
0          18h00  White shark                R. Collier, GSAF   
1  14h00  -15h00          NaN  K.McMurray, TrackingSharks.com   

                      pdf                                       href formula  \
0    2018.06.25-Wolfe.pdf  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  2018.06.18-McNeely.pdf  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   

  Case Number.2 Unnamed: 22 Unnamed: 23  
0    2018.06.25         NaN         NaN  
1    2018.06.18         NaN         NaN  

[2 rows x 22 columns]

Renombro COLUMNAS. Elimino mayúsculas y espacios.

Nota; más adelante me di cuenta que tendría que renombrar de nuevo la columna "fata(y/n)" para poder trabajar sobre ella. Haré referncia a ello más adelante.

In [66]:
attacks_copia = attacks.copy()

In [67]:
# Saco los 5 primeros elementos.
columnas = list(attacks.columns)
columnas[:5]

['Case Number', 'Date', 'Year', 'Type', 'Country']

In [68]:
diccio_nuevas = {columna: columna.replace(" ", "_").lower() for columna in columnas}

In [69]:
diccio_nuevas

{'Case Number': 'case_number',
 'Date': 'date',
 'Year': 'year',
 'Type': 'type',
 'Country': 'country',
 'Area': 'area',
 'Location': 'location',
 'Activity': 'activity',
 'Name': 'name',
 'Sex ': 'sex_',
 'Age': 'age',
 'Injury': 'injury',
 'Fatal (Y/N)': 'fatal_(y/n)',
 'Time': 'time',
 'Species ': 'species_',
 'Investigator or Source': 'investigator_or_source',
 'pdf': 'pdf',
 'href formula': 'href_formula',
 'href': 'href',
 'Case Number.1': 'case_number.1',
 'Case Number.2': 'case_number.2',
 'original order': 'original_order',
 'Unnamed: 22': 'unnamed:_22',
 'Unnamed: 23': 'unnamed:_23'}

In [70]:
attacks_renombrado = attacks.rename(columns=diccio_nuevas)
attacks_renombrado.sample()

case_number date  year type country area location activity name sex_  \
14089         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

       ... species_ investigator_or_source  pdf href_formula href  \
14089  ...      NaN                    NaN  NaN          NaN  NaN   

      case_number.1 case_number.2 original_order unnamed:_22 unnamed:_23  
14089           NaN           NaN            NaN         NaN         NaN  

[1 rows x 24 columns]

In [71]:
attacks.rename(columns=diccio_nuevas, inplace=True)

In [72]:
attacks.head(2)

case_number         date    year        type country        area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating     USA  California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked     USA     Georgia   

                         location  activity             name sex_  ...  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely     F  ...   

      species_          investigator_or_source                     pdf  \
0  White shark                R. Collier, GSAF    2018.06.25-Wolfe.pdf   
1          NaN  K.McMurray, TrackingSharks.com  2018.06.18-McNeely.pdf   

                                        href_formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href case_number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   

  case_number.2 original_order unnamed:_22 unnamed:_23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  

[2 rows x 24 columns]

Compruebo los datos que no son Nan en la columna unnamed:_23 para ver si hay algún dato relevante.

In [73]:
attacks[attacks["unnamed:_23"].notna()]

case_number                  date    year        type  \
4415    1952.03.30           30-Mar-1952  1952.0  Unprovoked   
5840  1878.09.14.R  Reported 14-Sep-1878  1878.0    Provoked   

                   country         area                    location  \
4415  NETHERLANDS ANTILLES      Curacao                         NaN   
5840                   USA  Connecticut  Branford, New Haven County   

                                             activity              name sex_  \
4415  Went to aid of child being menaced by the shark       A.J. Eggink    M   
5840                                          Fishing  Captain Pattison    M   

      ...                                           species_  \
4415  ...  Bull shark, 2.7 m [9'] was captured & dragged ...   
5840  ...                                                NaN   

                                 investigator_or_source  \
4415  J. Randall, p.352 in Sharks & Survival; H.D. B...   
5840                       St. Joseph Herald, 9/14/1878   

                            pdf  \
4415      1952.03.30-Eggink.pdf   
5840  1878.09.14.R-Pattison.pdf   

                                           href_formula  \
4415  http://sharkattackfile.net/spreadsheets/pdf_di...   
5840  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href case_number.1  \
4415  http://sharkattackfile.net/spreadsheets/pdf_di...    1952.03.30   
5840  http://sharkattackfile.net/spreadsheets/pdf_di...  1878.09.14.R   

     case_number.2 original_order unnamed:_22      unnamed:_23  
4415    1952.03.30         1888.0         NaN           Teramo  
5840  1878.09.14.R          463.0         NaN  change filename  

[2 rows x 24 columns]

Vuelvo a hacer un cambio de nombres en las columnas. Elimino los paréntesis de la columna "fatal" para poder utilizarla en la validación de hipótesis.

In [74]:
columnas_1 = list(attacks.columns)

In [75]:
diccio_nuevas_1 = {columna: columna.replace("(y/n)", "") for columna in columnas_1}

In [76]:
diccio_nuevas_1

{'case_number': 'case_number',
 'date': 'date',
 'year': 'year',
 'type': 'type',
 'country': 'country',
 'area': 'area',
 'location': 'location',
 'activity': 'activity',
 'name': 'name',
 'sex_': 'sex_',
 'age': 'age',
 'injury': 'injury',
 'fatal_(y/n)': 'fatal_',
 'time': 'time',
 'species_': 'species_',
 'investigator_or_source': 'investigator_or_source',
 'pdf': 'pdf',
 'href_formula': 'href_formula',
 'href': 'href',
 'case_number.1': 'case_number.1',
 'case_number.2': 'case_number.2',
 'original_order': 'original_order',
 'unnamed:_22': 'unnamed:_22',
 'unnamed:_23': 'unnamed:_23'}

In [77]:
attacks_renombrado = attacks.rename(columns=diccio_nuevas_1)
attacks_renombrado.sample()

case_number date  year type country area location activity name sex_  \
7364           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

      ... species_ investigator_or_source  pdf href_formula href  \
7364  ...      NaN                    NaN  NaN          NaN  NaN   

     case_number.1 case_number.2 original_order unnamed:_22 unnamed:_23  
7364           NaN           NaN            NaN         NaN         NaN  

[1 rows x 24 columns]

In [78]:
attacks.rename(columns=diccio_nuevas_1, inplace=True)

In [79]:
attacks.head(2)

case_number         date    year        type country        area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating     USA  California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked     USA     Georgia   

                         location  activity             name sex_  ...  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely     F  ...   

      species_          investigator_or_source                     pdf  \
0  White shark                R. Collier, GSAF    2018.06.25-Wolfe.pdf   
1          NaN  K.McMurray, TrackingSharks.com  2018.06.18-McNeely.pdf   

                                        href_formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href case_number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   

  case_number.2 original_order unnamed:_22 unnamed:_23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  

[2 rows x 24 columns]

Ahora ELIMINO COLUMNAS y FILAS que no me aportan valor.

1- COLUMNAS

Elimino COLUMNAS que no aportan nada por su significado; no aportan nada a la tabla.

In [80]:
attacks.drop(["pdf", "href_formula","href","unnamed:_22", "unnamed:_23"],axis=1,inplace=True)

In [81]:
attacks.head(2)

case_number         date    year        type country        area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating     USA  California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked     USA     Georgia   

                         location  activity             name sex_ age  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe    F  57   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely     F  11   

                                              injury fatal_           time  \
0  No injury to occupant, outrigger canoe and pad...      N          18h00   
1                         Minor injury to left thigh      N  14h00  -15h00   

      species_          investigator_or_source case_number.1 case_number.2  \
0  White shark                R. Collier, GSAF    2018.06.25    2018.06.25   
1          NaN  K.McMurray, TrackingSharks.com    2018.06.18    2018.06.18   

   original_order  
0          6303.0  
1          6302.0

Aquí elimino columnas que no aportan nada para mis hipótesis.

In [82]:
attacks.drop(["time", "species_","investigator_or_source","case_number.1", "case_number.2", "original_order"],axis=1,inplace=True)

In [83]:
attacks.head(2)

case_number         date    year        type country        area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating     USA  California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked     USA     Georgia   

                         location  activity             name sex_ age  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe    F  57   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely     F  11   

                                              injury fatal_  
0  No injury to occupant, outrigger canoe and pad...      N  
1                         Minor injury to left thigh      N

2-FILAS.

Elimino las filas en las que sus valores son todo NaN; no aporta nada a mi análisis. 

Para este paso he cambiado de data frame por si metía la pata. A apartir de aquí trabajo con el data frame "sharks"

In [84]:
sharks = attacks.dropna(how="all", axis=0)

In [85]:
sharks.shape

(8703, 13)

# LIMPIEZA por columnas.
### COLUMNA; date

Para la columna "date" me quedo solo con mes y año. El resto lo elimino. Al igual que antes no aporta valor pero lo dejo indicado.

In [86]:
sharks['newcol'] = sharks['date'].str.extract(r"(\w+-\d+)")

<ipython-input-86-65ad4d2e00fd>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sharks['newcol'] = sharks['date'].str.extract(r"(\w+-\d+)")


In [87]:
sharks.sample(2)

case_number  date    year        type    country             area  \
3046  1981.00.00.b  1981  1981.0  Unprovoked  AUSTRALIA  New South Wales   
6476             0   NaN     NaN         NaN        NaN              NaN   

       location     activity         name sex_  age                 injury  \
3046  Byron Bay  Surf-skiing  Warren Rowe    M   27  No injury, ski bitten   
6476        NaN          NaN          NaN  NaN  NaN                    NaN   

     fatal_ newcol  
3046      N    NaN  
6476    NaN    NaN

In [88]:
dates = list(sharks.date)
print(dates[0:5])

['25-Jun-2018', '18-Jun-2018', '09-Jun-2018', '08-Jun-2018', '04-Jun-2018']


In [89]:
newcoll= list(sharks.newcol)
print(newcoll[0:5])

['Jun-2018', 'Jun-2018', 'Jun-2018', 'Jun-2018', 'Jun-2018']


In [90]:
sharks.date=newcoll

/home/jesus_chao/anaconda3/envs/ironhack/lib/python3.8/site-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [91]:
# Compruebo que me ha cambiado las fechas.
print(sharks.date[0:5])

0    Jun-2018
1    Jun-2018
2    Jun-2018
3    Jun-2018
4    Jun-2018
Name: date, dtype: object


In [92]:
sharks.sample()

case_number      date    year        type country   area  \
2678  1989.07.07  Jul-1989  1989.0  Unprovoked     USA  Texas   

                             location  activity            name sex_ age  \
2678  Sargent Beach, Matagorda County  Playing   Rene Richbourg    F   9   

          injury fatal_    newcol  
2678  Leg bitten      N  Jul-1989

In [93]:
# Como he cambiado la date, puedo eliminar ya newcol
sharks.drop(["newcol"],axis=1, inplace=True)

/home/jesus_chao/anaconda3/envs/ironhack/lib/python3.8/site-packages/pandas/core/frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [94]:
sharks.sample()

case_number      date    year        type country  \
4231  1956.09.13  Sep-1956  1956.0  Unprovoked     NaN   

                                    area location               activity  \
4231  Near the Andaman & Nicobar Islands      NaN  Climbing back on ship   

      name sex_  age injury fatal_  
4231  male    M  NaN  FATAL      Y

#### COLUMNA; year

Aquí lo que hago es eliminar el ".0" de los años.

In [95]:
"""def year_(x):
    if ".0" in str(x):
        return str(x).replace(".0", "")
    else:
        return x"""

'def year_(x):\n    if ".0" in str(x):\n        return str(x).replace(".0", "")\n    else:\n        return x'

In [96]:
year_

<function src.cleaning.year_(x)>

In [97]:
sharks["year"] = sharks.year.apply(year_)

<ipython-input-97-34703d824513>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sharks["year"] = sharks.year.apply(year_)


In [98]:
print(sharks.year)

0        2018
1        2018
2        2018
3        2018
4        2018
         ... 
8698      NaN
8699      NaN
8700      NaN
8701      NaN
25722     NaN
Name: year, Length: 8703, dtype: object


In [99]:
# Este es el listado final de todas mis años.
sharks.year.unique()

array(['2018', '2017', nan, '2016', '2015', '2014', '2013', '2012',
       '2011', '2010', '2009', '2008', '2007', '2006', '2005', '2004',
       '2003', '2002', '2001', '2000', '1999', '1998', '1997', '1996',
       '1995', '1984', '1994', '1993', '1992', '1991', '1990', '1989',
       '1969', '1988', '1987', '1986', '1985', '1983', '1982', '1981',
       '1980', '1979', '1978', '1977', '1976', '1975', '1974', '1973',
       '1972', '1971', '1970', '1968', '1967', '1966', '1965', '1964',
       '1963', '1962', '1961', '1960', '1959', '1958', '1957', '1956',
       '1955', '1954', '1953', '1952', '1951', '1950', '1949', '1948',
       '1848', '1947', '1946', '1945', '1944', '1943', '1942', '1941',
       '1940', '1939', '1938', '1937', '1936', '1935', '1934', '1933',
       '1932', '1931', '1930', '1929', '1928', '1927', '1926', '1925',
       '1924', '1923', '1922', '1921', '1920', '1919', '1918', '1917',
       '1916', '1915', '1914', '1913', '1912', '1911', '1910', '1909',
       '1

### COLUMNA type.

Cambios hechos en esta columna: Cambio nombre de "Unprovoked" por "Accidental".Además he metido "Boatomg" y "Boating" en una sola columna "Boat" 

In [100]:
sharks.type.unique()

array(['Boating', 'Unprovoked', 'Invalid', 'Provoked', 'Questionable',
       'Sea Disaster', nan, 'Boat', 'Boatomg'], dtype=object)

In [101]:
"""def type_(x):
    if "Unprovoked" in str(x):
        return x.replace("Unprovoked", "Accidental")
    else:
        return x"""

'def type_(x):\n    if "Unprovoked" in str(x):\n        return x.replace("Unprovoked", "Accidental")\n    else:\n        return x'

In [102]:
type_

<function src.cleaning.type_(x)>

In [103]:
sharks["type"] = sharks.type.apply(type_)

<ipython-input-103-e5a48b295325>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sharks["type"] = sharks.type.apply(type_)


In [104]:
multi_p = sharks[(sharks.type == "Boatomg")]

In [105]:
multi_p

case_number      date  year     type      country          area  \
1152  2009.03.01.b  Mar-2009  2009  Boatomg  NEW ZEALAND  North Island   

      location activity                                             name sex_  \
1152  Taranaki  Fishing  boat, occupants: Boyd Rutherford & Hamish Roper    M   

      age                                       injury fatal_  
1152  NaN  No injury to occupants, shark bit propeller      N

In [106]:
"""def type_1(x):
    if "Boatomg" in str(x):
        return x.replace("Boatomg", "Boat")
    else:
        return x"""

'def type_1(x):\n    if "Boatomg" in str(x):\n        return x.replace("Boatomg", "Boat")\n    else:\n        return x'

In [107]:
type_1

<function src.cleaning.type_1(x)>

In [108]:
sharks["type"] = sharks.type.apply(type_1)

<ipython-input-108-240476cb040b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sharks["type"] = sharks.type.apply(type_1)


In [109]:
"""def type_2(x):
    if "Invalid" in str(x):
        return x.replace("Invalid", "Injured/Fatal")
    else:
        return x"""

'def type_2(x):\n    if "Invalid" in str(x):\n        return x.replace("Invalid", "Injured/Fatal")\n    else:\n        return x'

In [110]:
type_2

<function src.cleaning.type_2(x)>

In [111]:
sharks["type"] = sharks.type.apply(type_2)

<ipython-input-111-0df61136e9dd>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sharks["type"] = sharks.type.apply(type_2)


In [112]:
"""def type_3(x):
    if "Boating" in str(x):
        return x.replace("Boating", "Boat")
    else:
        return x"""

'def type_3(x):\n    if "Boating" in str(x):\n        return x.replace("Boating", "Boat")\n    else:\n        return x'

In [113]:
type_3

<function src.cleaning.type_3(x)>

In [114]:
sharks["type"] = sharks.type.apply(type_3)

<ipython-input-114-75c94eba09d1>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sharks["type"] = sharks.type.apply(type_3)


In [115]:
sharks.type.unique()

array(['Boat', 'Accidental', 'Injured/Fatal', 'Provoked', 'Questionable',
       'Sea Disaster', nan], dtype=object)

### COLUMNA; Pais, área y localización

Área y localización no me afecta a mis hipótesis; no los toco aunque se puedan mejorar

CAMBIOS: He convertido las letras minusculas en mayúsculas y eliminado "?". He metido también "ENGLAND" en "UNITED KINGDOM"

In [116]:
countries = list(sharks.country.unique())

In [117]:
len(countries)

213

In [118]:
print(countries[0:20])

['USA', 'AUSTRALIA', 'MEXICO', 'BRAZIL', 'ENGLAND', 'SOUTH AFRICA', 'THAILAND', 'COSTA RICA', 'MALDIVES', 'BAHAMAS', 'NEW CALEDONIA', 'ECUADOR', 'MALAYSIA', 'LIBYA', nan, 'CUBA', 'MAURITIUS', 'NEW ZEALAND', 'SPAIN', 'SAMOA']


Ponemos todos los países en mayúsculas.

In [119]:
"""def upper_c(x):
    return str(x).upper()"""

'def upper_c(x):\n    return str(x).upper()'

In [120]:
upper_c

<function src.cleaning.upper_c(x)>

In [121]:
sharks["country"] = sharks.country.apply(upper_c)

<ipython-input-121-51f75b38b595>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sharks["country"] = sharks.country.apply(upper_c)


Reemplazar ENGLAND por UNITED KINGDOM

In [122]:
"""def england_(x):
    if "ENGLAND" in str(x):
        return x.replace("ENGLAND", "UNITED KINGDOM")
    else:
        return x"""

'def england_(x):\n    if "ENGLAND" in str(x):\n        return x.replace("ENGLAND", "UNITED KINGDOM")\n    else:\n        return x'

In [123]:
england_

<function src.cleaning.england_(x)>

In [124]:
sharks["country"] = sharks.country.apply(england_)

<ipython-input-124-9a5641fac9da>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sharks["country"] = sharks.country.apply(england_)


ELIMINO signo de interrogación.

In [125]:
"""def question_(x):
    if "?" in str(x):
        return x.replace("?", "")
    else:
        return x"""

'def question_(x):\n    if "?" in str(x):\n        return x.replace("?", "")\n    else:\n        return x'

In [126]:
question_

<function src.cleaning.question_(x)>

In [127]:
sharks["country"] = sharks.country.apply(question_)

<ipython-input-127-1d540e1a7bd3>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sharks["country"] = sharks.country.apply(question_)


In [128]:
# Este es el listado final de todos mis paises.
c_ =list(sharks.country.unique())

In [129]:
c_[0:5]

['USA', 'AUSTRALIA', 'MEXICO', 'BRAZIL', 'UNITED KINGDOM']

### COLUMNA; Activity

Aunque mis hipótesis afectan a actividades, para validarlo he optado por otra opción que no sea limpieza. Ver la validación. Se han creado columnas FALSE y TRUE.

In [130]:
activities = list(sharks.activity.unique())

In [131]:
activities[0:5]

['Paddling', 'Standing', 'Surfing', 'Free diving', 'Kite surfing']

COLUMNA: sex_

Limpio la columna para quedarme sólo con F, M o None

In [132]:
gender = list(sharks.sex_.unique())

In [133]:
gender

['F', 'M', nan, 'M ', 'lli', 'N', '.']

In [134]:
sharks.head(2)

case_number      date  year        type country        area  \
0  2018.06.25  Jun-2018  2018        Boat     USA  California   
1  2018.06.18  Jun-2018  2018  Accidental     USA     Georgia   

                         location  activity             name sex_ age  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe    F  57   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely     F  11   

                                              injury fatal_  
0  No injury to occupant, outrigger canoe and pad...      N  
1                         Minor injury to left thigh      N

In [135]:
"""def gender_(x):
    special_characters = 'FM'
    if str(x) in special_characters:
        return x
    else:
        return str(x).replace(str(x), "None")"""

'def gender_(x):\n    special_characters = \'FM\'\n    if str(x) in special_characters:\n        return x\n    else:\n        return str(x).replace(str(x), "None")'

In [136]:
gender_

<function src.cleaning.gender_(x)>

In [137]:
sharks["genero"] = sharks.sex_.apply(gender_)

<ipython-input-137-0220de8c0124>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sharks["genero"] = sharks.sex_.apply(gender_)


In [138]:
sharks.head()

case_number      date  year           type    country             area  \
0  2018.06.25  Jun-2018  2018           Boat        USA       California   
1  2018.06.18  Jun-2018  2018     Accidental        USA          Georgia   
2  2018.06.09  Jun-2018  2018  Injured/Fatal        USA           Hawaii   
3  2018.06.08  Jun-2018  2018     Accidental  AUSTRALIA  New South Wales   
4  2018.06.04  Jun-2018  2018       Provoked     MEXICO           Colima   

                         location     activity             name sex_  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

                                              injury fatal_ genero  
0  No injury to occupant, outrigger canoe and pad...      N      F  
1                         Minor injury to left thigh      N      F  
2       Injury to left lower leg from surfboard skeg      N      M  
3                          Minor injury to lower leg      N      M  
4  Lacerations to leg & hand shark PROVOKED INCIDENT      N      M

In [139]:
gender = list(sharks.genero.unique())

In [140]:
gender

['F', 'M', 'None']

In [141]:
sharks.drop(["sex_"],axis=1,inplace=True)

/home/jesus_chao/anaconda3/envs/ironhack/lib/python3.8/site-packages/pandas/core/frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [142]:
sharks.head(2)

case_number      date  year        type country        area  \
0  2018.06.25  Jun-2018  2018        Boat     USA  California   
1  2018.06.18  Jun-2018  2018  Accidental     USA     Georgia   

                         location  activity             name age  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe  57   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely   11   

                                              injury fatal_ genero  
0  No injury to occupant, outrigger canoe and pad...      N      F  
1                         Minor injury to left thigh      N      F

### COLUMNA: fatal_

Limpio para quedarme con N, Y o None

In [143]:
fatal = list(sharks.fatal_.unique())

In [144]:
fatal

['N', 'Y', nan, 'M', 'UNKNOWN', '2017', ' N', 'N ', 'y']

In [145]:
"""def fatal_(x):
    special_characters = 'NYy'
    if str(x) in special_characters:
        return x.upper()
    else:
        return str(x).replace(str(x), "None")"""

'def fatal_(x):\n    special_characters = \'NYy\'\n    if str(x) in special_characters:\n        return x.upper()\n    else:\n        return str(x).replace(str(x), "None")'

In [146]:
fatal_

<function src.cleaning.fatal_(x)>

In [147]:
sharks["fatal"] = sharks.fatal_.apply(fatal_)

<ipython-input-147-988055907673>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sharks["fatal"] = sharks.fatal_.apply(fatal_)


In [148]:
sharks.head(2)

case_number      date  year        type country        area  \
0  2018.06.25  Jun-2018  2018        Boat     USA  California   
1  2018.06.18  Jun-2018  2018  Accidental     USA     Georgia   

                         location  activity             name age  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe  57   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely   11   

                                              injury fatal_ genero fatal  
0  No injury to occupant, outrigger canoe and pad...      N      F     N  
1                         Minor injury to left thigh      N      F     N

In [149]:
fatal_ = list(sharks.fatal.unique())

In [150]:
fatal_

['N', 'Y', 'None']

In [151]:
sharks.drop(["fatal_"],axis=1,inplace=True)

/home/jesus_chao/anaconda3/envs/ironhack/lib/python3.8/site-packages/pandas/core/frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [152]:
sharks.head(2)

case_number      date  year        type country        area  \
0  2018.06.25  Jun-2018  2018        Boat     USA  California   
1  2018.06.18  Jun-2018  2018  Accidental     USA     Georgia   

                         location  activity             name age  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe  57   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely   11   

                                              injury genero fatal  
0  No injury to occupant, outrigger canoe and pad...      F     N  
1                         Minor injury to left thigh      F     N

# Validación HIPOTESIS

### 1- USA es donde más accidentes de tiburón se producen.

Cuento los valores de la columna para validar hipótesis.

In [153]:
sharks.country.value_counts()

NAN                      2451
USA                      2229
AUSTRALIA                1338
SOUTH AFRICA              579
PAPUA NEW GUINEA          134
                         ... 
DIEGO GARCIA                1
ARGENTINA                   1
THE BALKANS                 1
NORTH ATLANTIC OCEAN        1
CEYLON (SRI LANKA)          1
Name: country, Length: 206, dtype: int64

SE CUMPLE.

### 2- La mayor parte de los accidentes son no provocados.

Cuento los valores de la columna para validar.

In [154]:
sharks.type.value_counts()

Accidental       4595
Provoked          574
Injured/Fatal     547
Boat              341
Sea Disaster      239
Questionable        2
Name: type, dtype: int64

SE CUMPLE.

### 3- Fishing y Surfing; mayores causas de accidentes.

 Creo dos columnas Tue or False para datos que terminan en "ishing" y "urfing". Los cuento y cuento los fishing y surfing.

In [155]:
sharks["fishing"] = sharks["activity"].str.contains("ishing")

<ipython-input-155-bfcf38036315>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sharks["fishing"] = sharks["activity"].str.contains("ishing")


In [156]:
sharks["surfing"] = sharks["activity"].str.contains("urfing")

<ipython-input-156-d92449e05581>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sharks["surfing"] = sharks["activity"].str.contains("urfing")


In [157]:
sharks.head(2)

case_number      date  year        type country        area  \
0  2018.06.25  Jun-2018  2018        Boat     USA  California   
1  2018.06.18  Jun-2018  2018  Accidental     USA     Georgia   

                         location  activity             name age  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe  57   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely   11   

                                              injury genero fatal fishing  \
0  No injury to occupant, outrigger canoe and pad...      F     N   False   
1                         Minor injury to left thigh      F     N   False   

  surfing  
0   False  
1   False

In [158]:
sharks.fishing.value_counts()

False    4586
True     1172
Name: fishing, dtype: int64

In [159]:
sharks.surfing.value_counts()

False    4630
True     1128
Name: surfing, dtype: int64

In [160]:
sharks.isnull().sum()

case_number       1
date           2946
year           2403
type           2405
country           0
area           2856
location       2941
activity       2945
name           2611
age            5232
injury         2429
genero            0
fatal             0
fishing        2945
surfing        2945
dtype: int64

2945 nulls

Calculo el porcentaje de personas con accidente causado por surfing y fishing. Si mayor de 50% se valida.

In [161]:
suma_fish_surf = 1128 + 1172

In [162]:
Porcentaje = suma_fish_surf/(4630+1128)

In [163]:
Porcentaje

0.399444251476207

NO SE CUMPLE.

### 4- Mueren más hombres que mujeres.

Cuento los valores de la columna para validar.

In [164]:
hombres_muertos = sharks[(sharks.genero == "M") & (sharks.fatal == "Y")]

In [165]:
hombres_muertos.genero.value_counts()

M    1173
Name: genero, dtype: int64

In [166]:
mujeres_muertas = sharks[(sharks.genero == "F") & (sharks.fatal == "Y")]

In [167]:
mujeres_muertas.genero.value_counts()

F    107
Name: genero, dtype: int64

SE CUMPLE.

# Conclusión

Las HIPOTESIS 1, 2, y 4 son verdaderas la 3 es falsa.

Esta es el aspecto final del csv

In [168]:
sharks.head(3)

case_number      date  year           type country        area  \
0  2018.06.25  Jun-2018  2018           Boat     USA  California   
1  2018.06.18  Jun-2018  2018     Accidental     USA     Georgia   
2  2018.06.09  Jun-2018  2018  Injured/Fatal     USA      Hawaii   

                         location  activity             name age  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe  57   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely   11   
2                    Habush, Oahu   Surfing      John Denges  48   

                                              injury genero fatal fishing  \
0  No injury to occupant, outrigger canoe and pad...      F     N   False   
1                         Minor injury to left thigh      F     N   False   
2       Injury to left lower leg from surfboard skeg      M     N   False   

  surfing  
0   False  
1   False  
2    True

Exporto CSV limpio.

In [169]:
sharks.to_csv("sharks.csv", index = False)